src/node/historical_queries_utils.cpp

```diff
@@ -145,5 +292,54 @@ namespace ccf
 
       return true;
     }
+
+    bool populate_cose_service_endorsements(
+      ccf::kv::ReadOnlyTx& tx,
+      ccf::historical::StatePtr& state,
+      AbstractStateCache& state_cache)
+    {
+      const auto service_info = tx.template ro<Service>(Tables::SERVICE)->get();
+      const auto service_start = service_info->current_service_create_txid;
+      if (!service_start)
+      {
+        LOG_FAIL_FMT("Service start txid not found");
+        assert(false); // In debug, fail fast.
+
+        return true;
+      }
+
+      const auto target_seq = state->transaction_id.seqno;
+      if (service_start->seqno <= target_seq)
+      {
+        LOG_TRACE_FMT(
+          "Target seqno {} belongs to current service started at {}",
+          target_seq,
+          service_start->seqno);
+        return true;
+      }
+
+      try
+      {
+        auto result =
+          fetch_endorsements_for(tx, state_cache, state->transaction_id.seqno);
+        if (!result.endorsements)
+        {
+          const bool final_result = !result.retry;
+          return final_result;
+        }
+        state->receipt->cose_endorsements = result.endorsements.value();
+      }
+      catch (const std::logic_error& e)
+      {
+        LOG_FAIL_FMT("Failed to fetch endorsements: {}", e.what());
+
+        assert(false); // In debug, fail fast.
+
+        return true; // In release, return true to avoid being stuck waiting for
```

Comment:
why not throw?

Reply:
Regular service endorsements don't throw, this doesn't either.

In this case a receipt with no endorsement will be provided, alternatively we could've fail a whole receipt creation.

I don't know what's a best choice here.

Reply:
Decided to throw an catch in adapter code